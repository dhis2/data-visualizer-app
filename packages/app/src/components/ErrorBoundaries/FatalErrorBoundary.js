import React, { Component } from 'react';
import PropTypes from 'prop-types';
import i18n from '@dhis2/d2-i18n';
import { withStyles } from '@material-ui/core/styles';
import InfoIcon from '@material-ui/icons/InfoOutlined';
import styles from './styles/FatalErrorBoundary.style';

const translatedErrorHeading = i18n.t(
    'An error occurred in the DHIS2 Data Visualizer application.'
);

const replaceNewlinesWithBreaks = text =>
    text
        .split('\n')
        .reduce((out, line, i) => [...out, line, <br key={i} />], []);

export class FatalErrorBoundary extends Component {
    constructor(props) {
        super(props);
        this.state = {
            error: null,
            errorInfo: null,
            drawerOpen: false,
        };
    }

    componentDidCatch(error, errorInfo) {
        this.setState({
            error,
            errorInfo,
        });
    }

    toggleTechInfoDrawer = () => {
        this.setState({
            drawerOpen: !this.state.drawerOpen,
        });
    };

    render() {
        const { classes, children } = this.props;
        if (this.state.error) {
            return (
                <div className={classes.mask}>
                    <div className={classes.container}>
                        <InfoIcon className={classes.icon} />
                        <div className={classes.message}>
                            {i18n.t('Something went wrong')}
                        </div>
                        <div
                            className={classes.link}
                            onClick={() => window.location.reload()}
                        >
                            {i18n.t('Refresh to try again')}
                        </div>
                        <div
                            className={classes.drawerToggle}
                            onClick={this.toggleTechInfoDrawer}
                        >
                            {this.state.drawerOpen
                                ? i18n.t('Hide technical details')
                                : i18n.t('Show technical details')}
                        </div>
                        <div
                            className={
                                this.state.drawerOpen
                                    ? classes.drawerVisible
                                    : classes.drawerHidden
                            }
                        >
                            <div className={classes.errorIntro}>
                                {translatedErrorHeading}
                                <br />
                                {i18n.t(
                                    'The following information may be requested by technical support.'
                                )}
                            </div>
                            <div className={classes.errorDetails}>
                                {[
                                    replaceNewlinesWithBreaks(
                                        this.state.error.stack +
                                            '\n---' +
                                            this.state.errorInfo.componentStack
                                    ),
                                ]}
                            </div>
                        </div>
                    </div>
                </div>
            );
        }

        return children;
    }
}

FatalErrorBoundary.propTypes = {
    classes: PropTypes.object.isRequired,
    children: PropTypes.node.isRequired,
};

export default withStyles(styles)(FatalErrorBoundary);
