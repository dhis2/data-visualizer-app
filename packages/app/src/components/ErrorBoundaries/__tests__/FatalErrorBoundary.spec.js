import React from 'react'
import { FatalErrorBoundary } from '../FatalErrorBoundary'
import { shallow } from 'enzyme'

const Something = () => {
    // Placeholder
    return null
}

describe('FatalErrorBoundary', () => {
    it('Should render the normal tree when no error occurs', () => {
        const wrapper = shallow(
            <FatalErrorBoundary classes={{}}>
                <div>
                    <span id="testme">Hello there!</span>
                </div>
            </FatalErrorBoundary>
        )

        expect(wrapper.find('span#testme').length).toBe(1)
    })

    it('Should render the error mask when an error is thrown', () => {
        const wrapper = shallow(
            <FatalErrorBoundary classes={{}}>
                <Something />
            </FatalErrorBoundary>
        )

        expect(wrapper.find(Something).length).toBe(1)

        const testErrorText =
            'This is a test of the emergency alert system.  This is only a test.'
        wrapper.find(Something).simulateError(new Error(testErrorText))
        expect(wrapper.contains('Something went wrong')).toBe(true)
    })

    it('Should match the snapshot tree when error is manually invoked', () => {
        const wrapper = shallow(
            <FatalErrorBoundary classes={{}}>
                <Something />
            </FatalErrorBoundary>
        )

        const error = {
            stack: 'This is the stack',
        }
        const errorInfo = {
            componentStack: 'This is the component stack',
        }

        wrapper.setState({ error, errorInfo })
        expect(wrapper.contains('Something went wrong')).toBe(true)

        expect(wrapper).toMatchSnapshot()
    })
})
