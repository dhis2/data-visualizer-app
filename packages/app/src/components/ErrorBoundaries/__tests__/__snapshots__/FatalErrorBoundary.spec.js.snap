// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatalErrorBoundary Should match the snapshot tree when error is manually invoked 1`] = `
<div>
  <div>
    <pure(InfoOutlinedIcon) />
    <div>
      Something went wrong
    </div>
    <div
      onClick={[Function]}
    >
      Refresh to try again
    </div>
    <div
      onClick={[Function]}
    >
      Show technical details
    </div>
    <div>
      <div>
        An error occurred in the DHIS2 Data Visualizer application.
        <br />
        The following information may be requested by technical support.
      </div>
      <div>
        This is the stack
        <br
          key="0"
        />
        ---This is the component stack
        <br
          key="1"
        />
      </div>
    </div>
  </div>
</div>
`;
